### PR TITLE
Move splitting primitives to serialize library

### PR DESCRIPTION
**Pull request by saulshanabrook**
This PR moves the ability to split primitive nodes into the serialize library from this library and also extends it to support splitting arbitrary nodes. See https://github.com/egraphs-good/egraph-serialize/pull/14 for the upstream PR that corresponds to this one, which includes an example using this new behavior.

By default, this should result in the same behavior as before, but just with the ability to split more e-classes besides just primitive nodes if desired.

In order to move the splitting logic upstream, I exposed a way to check if a serialized node ID is a primitive. In a way, this extends https://github.com/egraphs-good/egglog/pull/396 by also now supporting public deserialization/serialization for serialized node IDs as well as class IDs. This should also be useful if you do extraction externally and then want to convert back to egglog values.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1552,24 +1553,6 @@ impl EGraph {
...
-        max_functions: usize,
-        max_calls_per_function: usize,
-    ) -> egraph_serialize::EGraph {
-        let config = SerializeConfig {
-            split_primitive_outputs,
-            max_functions: Some(max_functions),
-            max_calls_per_function: Some(max_calls_per_function),
-            ..Default::default()
-        };
-        self.serialize(config)
```
Comment (saulshanabrook) [inline src/lib.rs:1570]: I removed this help function since it wasn't really doing much

File: src/main.rs [hunk 2]
```diff
@@ -43,6 +43,9 @@ struct Args {
     /// Maximum number of calls per function to render in dot/svg output
     #[clap(long, default_value = "40")]
     max_calls_per_function: usize,
+    /// Number of times to inline leaves
+    #[clap(long, default_value = "0")]
+    serialize_n_inline_leaves: usize,
```
Comment (saulshanabrook) [inline src/main.rs:48]: I added the ability to inline leaves via the egglog CLI when making visualizations/json

File: src/main.rs [hunk 3]
```diff
@@ -209,31 +212,25 @@ fn main() {
...
+
+            // if we are splitting primitive outputs, add `-split` to the end of the file name
+            let serialize_filename = if args.serialize_split_primitive_outputs {
+                input.with_file_name(format!(
+                    "{}-split",
+                    input.file_stem().unwrap().to_str().unwrap()
+                ))
+            } else {
+                input.clone()
+            };
```
Comment (saulshanabrook) [inline src/main.rs:233]: I moved all this processing into one branch, so it would only be serialized once for JSON or for visualization.

File: src/sort/set.rs [hunk 4]
```diff
@@ -190,6 +190,11 @@ impl Sort for SetSort {
         }
         Some((cost, expr))
     }
+
+    fn serialized_name(&self, _value: &Value) -> Symbol {
+        "set-of".into()
+    }
+
```
Comment (saulshanabrook) [inline src/sort/set.rs:197]: This is an unrelated change... It means that when set constructors are serialized they will use this as their name instead of the name of the sort.

File: src/main.rs [hunk 5]
```diff
@@ -1,5 +1,5 @@
 use clap::Parser;
-use egglog::{EGraph, Error, RunMode, SerializeConfig};
+use egglog::{EGraph, Error, RunMode, SerializeConfig, SerializedNode};
```
Comment (saulshanabrook) [inline src/main.rs:2]:
```suggestion
use egglog::{EGraph, Error, RunMode, SerializeConfig};
```

**Review by yihozhang — APPROVED** [review]
